modules/public/public.go

```diff
@@ -5,86 +5,45 @@
 package public
 
 import (
-	"log"
 	"net/http"
-	"path"
+	"os"
 	"path/filepath"
 	"strings"
 
 	"code.gitea.io/gitea/modules/httpcache"
+	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
 )
 
 // Options represents the available options to configure the handler.
 type Options struct {
-	Directory   string
-	IndexFile   string
-	SkipLogging bool
-	FileSystem  http.FileSystem
-	Prefix      string
+	Directory string
+	Prefix    string
 }
 
-// KnownPublicEntries list all direct children in the `public` directory
-var KnownPublicEntries = []string{
-	"css",
-	"fonts",
-	"img",
-	"js",
-	"serviceworker.js",
-	"vendor",
-}
-
-// Custom implements the static handler for serving custom assets.
-func Custom(opts *Options) func(next http.Handler) http.Handler {
-	return opts.staticHandler(path.Join(setting.CustomPath, "public"))
-}
-
-// staticFileSystem implements http.FileSystem interface.
-type staticFileSystem struct {
-	dir *http.Dir
-}
-
-func newStaticFileSystem(directory string) staticFileSystem {
-	if !filepath.IsAbs(directory) {
-		directory = filepath.Join(setting.AppWorkPath, directory)
+// AssetsHandler implements the static handler for serving custom or original assets.
+func AssetsHandler(opts *Options) func(resp http.ResponseWriter, req *http.Request) {
+	var custPath = filepath.Join(setting.CustomPath, "public")
+	if !filepath.IsAbs(custPath) {
+		custPath = filepath.Join(setting.AppWorkPath, custPath)
 	}
-	dir := http.Dir(directory)
-	return staticFileSystem{&dir}
-}
 
-func (fs staticFileSystem) Open(name string) (http.File, error) {
-	return fs.dir.Open(name)
-}
-
-// StaticHandler sets up a new middleware for serving static files in the
-func StaticHandler(dir string, opts *Options) func(next http.Handler) http.Handler {
-	return opts.staticHandler(dir)
-}
+	if !filepath.IsAbs(opts.Directory) {
+		opts.Directory = filepath.Join(setting.AppWorkPath, opts.Directory)
+	}
 
-func (opts *Options) staticHandler(dir string) func(next http.Handler) http.Handler {
-	return func(next http.Handler) http.Handler {
-		// Defaults
-		if len(opts.IndexFile) == 0 {
-			opts.IndexFile = "index.html"
-		}
-		// Normalize the prefix if provided
-		if opts.Prefix != "" {
-			// Ensure we have a leading '/'
-			if opts.Prefix[0] != '/' {
-				opts.Prefix = "/" + opts.Prefix
-			}
-			// Remove any trailing '/'
-			opts.Prefix = strings.TrimRight(opts.Prefix, "/")
+	return func(resp http.ResponseWriter, req *http.Request) {
+		// custom files
+		if opts.handle(resp, req, http.Dir(custPath), opts.Prefix) {
+			return
 		}
-		if opts.FileSystem == nil {
-			opts.FileSystem = newStaticFileSystem(dir)
+
+		// internal files
+		if opts.handle(resp, req, fileSystem(opts.Directory), opts.Prefix) {
+			return
 		}
 
-		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-			if !opts.handle(w, req, opts) {
-				next.ServeHTTP(w, req)
-			}
-		})
+		resp.WriteHeader(404)
```

Comment:
use `http.StatusNotFound`

```suggestion
		resp.WriteHeader(http.StatusNotFound)
```